### ISSUE TITLE
Init

### ISSUE THREAD

**Issue by HenryTheHamster** [HenryTheHamster/goalie#1]
You are an AI coding agent. Implement an end-to-end, reproducible AI/ML prediction framework for
football (soccer) goals, focused on:
- Leagues: English Premier League (EPL) and English Championship
- Targets: (A) home goals and away goals distributions, (B) derived Over/Under probabilities
- Timing: rolling pre-match predictions at fixed snapshot cutoffs: T-7d, T-3d, T-24h, T-6h, T-1h
- Features: layered team strength signals (Elo-style attack/defence + EWMA rolling stats + rolling averages)
- Models: statistical baseline + ML challenger + Bayesian model
- Evaluation: likelihood/fit, calibration, market utility metrics, and P&L backtests (P&L is diagnostic only)
- Storage: Parquet snapshots + DuckDB query layer + per-run Markdown report + run manifest
- Data source: API-Football (DIRECT auth, not RapidAPI), mirrored locally as the canonical raw dataset

HARD RULES (must enforce):
1) NO TARGET LEAKAGE
   - No feature may use post-cutoff information relative to a snapshot time.
   - Odds data must NOT be used unless the feature flag `--feature-group odds` is explicitly enabled.
   - If leakage is detected or suspected, the run must fail with a clear error.
2) INTERPRETABILITY FIRST
   - Every feature must have a stable name, a human-readable description, and be logged in a feature dictionary.
   - The pipeline must support feature-group ablation (enable/disable feature blocks).
3) API-FOOTBALL DATA QUALITY RULES (must enforce):
   - Training uses only final results: status in {FT, AET, PEN}.
   - Postponed/abandoned/suspended/etc are excluded from training but kept for fixture tracking.
   - Deduplicate/reconcile by fixture_id; raw versions are append-only; one canonical row per fixture_id.
   - Enforce UTC timestamps and league/season consistency based on API parameters.

DELIVERY STYLE:
- Build in incremental vertical slices. Each slice must end with runnable CLI commands + metrics output.
- Prefer simple, reliable engineering over cleverness.
- Use Python. Keep dependencies minimal and well-known.
- Provide a clean repo structure, runnable locally.

REPO STRUCTURE (create these paths):
- README.md
- pyproject.toml (or requirements.txt)
- src/
  - config/
  - ingest/
  - data/
  - features/
  - models/
  - eval/
  - reports/
  - cli.py
- scripts/
- tests/
- artifacts/
  - parquet/
    - raw/
      - api_football/
        - fixtures/
        - teams/
        - leagues/
    - canonical/
      - matches/
      - fixtures/
      - teams/
      - leagues/
    - features/
    - predictions/
    - metrics/
  - runs/
    - <run_id>/
      - manifest.json
      - report.md
- duckdb/
  - football.duckdb

API-FOOTBALL INTEGRATION REQUIREMENTS (DIRECT AUTH):
- Config via env vars:
  - APIFOOTBALL_API_KEY=<key>
  - APIFOOTBALL_BASE_URL=<default to official API-Football v3 base>
- Implement a small HTTP client with:
  - timeouts, retries with backoff, and rate-limit friendly pacing
  - request/response logging (without leaking the API key)
  - local response caching OPTIONAL, but local raw parquet mirror is required

Endpoints to implement first (Option 2: fixtures + teams + leagues):
- Fixtures:
  - GET /fixtures?league={league_id}&season={season}
  - GET /fixtures?league={league_id}&season={season}&from={YYYY-MM-DD}&to={YYYY-MM-DD} (optional for incremental)
  - Must capture fixture.id, fixture.date (UTC), fixture.status.* , teams.*, goals.*, score.*
- Teams:
  - GET /teams?league={league_id}&season={season} (or equivalent)
  - Must capture team.id, team.name, and stable identifiers
- Leagues:
  - GET /leagues?id={league_id} (or a leagues listing filtered to required)
  - Must capture league.id, league.name, country, seasons list

IMPORTANT: Create an internal canonical schema regardless of API shape:
- matches (training-ready, final results only):
  - match_id (string or int) = fixture.id
  - league (string: "EPL"|"CHAMP")
  - league_id (int)
  - season (int)
  - date_time_utc (timestamp)
  - home_team_id (int), away_team_id (int)
  - home_team (string), away_team (string)
  - home_goals (int), away_goals (int)
  - status_short (string)
  - last_updated_utc (timestamp)
  - source_version_id (string)  # hash of raw payload or ingestion batch id
- fixtures (upcoming + all statuses):
  - match_id, league, league_id, season, date_time_utc, team ids/names, status_short, last_updated_utc, source_version_id
- teams:
  - team_id, team_name, league_id, season, last_updated_utc, source_version_id
- leagues:
  - league_id, league_name, country, season, last_updated_utc, source_version_id

MIRRORING (local raw source-of-truth):
- Implement CLI: `ingest api-football sync`
  - Required args:
    - --league <EPL|CHAMP>
    - --league-id <int> (configurable mapping)
    - --season <int>
  - Behavior:
    1) Call leagues/teams/fixtures endpoints for that league+season
    2) Write RAW parquet append-only under:
       artifacts/parquet/raw/api_football/{leagues|teams|fixtures}/
       partitioned by league, league_id, season, ingest_date, batch_id
    3) Build/refresh CANONICAL parquet under:
       artifacts/parquet/canonical/{leagues|teams|fixtures|matches}/
       partitioned by league, season
    4) Update DuckDB tables/views to point at canonical parquet

- Implement canonicalization rules:
  - Deduplicate by match_id using latest last_updated_utc (or latest batch if last_updated missing)
  - Keep raw versions append-only; never overwrite raw
  - For matches (training):
    - include only status_short in {FT, AET, PEN}
    - require home_goals and away_goals non-null ints

DATA ASSUMPTIONS (non-API fallback for tests):
- Include a synthetic/demo dataset generator and a CSV loader producing the canonical schema.

CORE CONCEPTS TO IMPLEMENT:

A) SNAPSHOT ENGINE
- A run produces predictions for upcoming fixtures at multiple cutoffs (T-7d, T-3d, T-24h, T-6h, T-1h).
- Each snapshot is a “point-in-time” view with a strict cutoff timestamp.
- Store each snapshot’s:
  - features parquet
  - predictions parquet
  - metrics parquet
- Include in manifest.json:
  - run_id, git hash (if available), timestamp
  - canonical dataset paths + counts
  - snapshot cutoffs and counts
  - model configs and hyperparameters
  - enabled feature groups
  - evaluation summary

B) FEATURE PIPELINE (feature groups + ablations)
Implement feature groups with explicit toggles:
- base:
  - home advantage indicator
  - season / league identifiers
- rolling_avgs:
  - rolling GF/GA for home/away teams over last N matches (configurable N)
- ewma_form:
  - EWMA of GF/GA and goal difference (configurable alpha)
- elo_attack_defence:
  - Elo-like rating system with separate attack and defence parameters
  - updates must obey cutoff timestamps
- (optional, disabled by default) odds:
  - placeholder scaffolding only; do not use unless explicitly enabled

Each feature must have:
- name, group, description, data type
- generation code that takes (matches_history <= cutoff, fixtures_to_predict, config)

C) MODELS
Implement three model families, all producing:
- lambda_home, lambda_away (expected goals)
- a goal probability matrix P(home_goals=i, away_goals=j) for i,j up to max_goals (configurable, default 8)
- derived markets:
  - P(total_goals > line) for common lines (1.5, 2.5, 3.5) configurable
  - optional: P(BTTS) derived from matrix

Model 1 (Stat baseline):
- Dixon–Coles style / bivariate Poisson baseline driven by team attack/defence + home advantage
- If full Dixon–Coles is too heavy, implement independent Poisson as v1 but keep the interface identical

Model 2 (ML challenger):
- Predict lambda_home and lambda_away using a standard regression model (sklearn HistGradientBoostingRegressor preferred)
- Must use the same feature table and snapshot discipline
- Provide feature importance output (where supported)

Model 3 (Bayesian):
- Hierarchical Poisson model for team attack/defence with uncertainty
- If heavy probabilistic libs are avoided, implement a lightweight approximation; keep outputs compatible
- Output uncertainty summaries if feasible (e.g., intervals on lambdas)

D) TRAINING & EVALUATION
Splits:
- Primary: rolling-origin / walk-forward evaluation (expanding window)
- Secondary: strict temporal holdout split

Metrics:
1) Fit quality: log-likelihood / negative log-likelihood on observed goals (+ optional MAE/RMSE)
2) Calibration: reliability summaries for O/U 2.5 (and others) + Brier scores
3) Market utility (no betting by default): sharpness/stability by cutoff (placeholders ok)
4) P&L backtests (diagnostic only, behind a flag): simple conservative strategy; DO NOT tune to P&L

E) STORAGE LAYER
Parquet:
- Write outputs into artifacts/parquet/{features,predictions,metrics}/ partitioned by:
  league, season, snapshot_cutoff, run_id, model
DuckDB:
- Build/update duckdb/football.duckdb with tables/views:
  - canonical_matches, canonical_fixtures, canonical_teams, canonical_leagues
  - features, predictions, metrics
- Provide helper SQL views for analysis (per league, per cutoff, per model)

F) REPORTING
Generate artifacts/runs/<run_id>/report.md containing:
- run metadata + enabled feature groups
- data quality report:
  - counts by fixture status
  - excluded rows and reasons
  - duplicates/changes detected
- dataset summary
- evaluation summary tables by league and cutoff
- calibration summaries
- model comparison
- notes on leakage checks performed

G) CLI
Implement CLI entrypoint with subcommands (argparse or typer):
- ingest api-football sync --league ... --league-id ... --season ...
- canonicalize --league ... --season ...
- build-features --cutoffs "7d,3d,24h,6h,1h" --asof <datetime> --league ...
- train --model <stat|ml|bayes> --split <rolling|holdout> ...
- predict --model ... --cutoffs ...
- eval --model ... --split ...
- run-all --league ... --seasons ... --cutoffs ... --models "stat,ml,bayes" --split rolling
- report --run-id <id>

Each command must:
- be runnable independently
- write outputs to the contract paths
- log key info to console

LEAKAGE PREVENTION MECHANISMS (must implement):
- Central “asof/cutoff” timestamp carried through every pipeline stage.
- Assertion helpers that verify:
  - no rows in history have timestamp > cutoff
  - no features reference future matches relative to fixture time and cutoff
  - rolling windows use only matches strictly before the fixture time and <= cutoff
- Unit tests:
  - at least one test that intentionally introduces future data and expects failure.

IMPLEMENTATION PLAN (must follow vertical slices):
Slice 1: Repo skeleton + demo dataset + basic CLI + parquet write + duckdb load
Slice 2: API-Football client + `ingest api-football sync` + raw mirror + canonicalization + duckdb tables/views
Slice 3: Snapshot engine + base features + stat baseline (independent Poisson ok) + predict + eval (NLL)
Slice 4: Team strength features (rolling + EWMA + Elo A/D) + ablation toggles + expanded eval + data quality report
Slice 5: ML model + comparison harness + feature importance output
Slice 6: Bayesian model + uncertainty outputs (if feasible)
Slice 7: Reporting (markdown) + run manifest + tests + polishing

ACCEPTANCE CRITERIA:
- `ingest api-football sync` produces raw parquet, canonical parquet, and updates DuckDB
- `run-all` completes and produces:
  - parquet features/predictions/metrics
  - updated DuckDB tables/views
  - run manifest.json
  - report.md including data quality section
- Evaluation includes results per league and cutoff for all models
- Feature groups can be ablated via CLI flags
- Leakage checks exist and at least one test proves they work

Now implement this framework.
Start with Slice 1 and proceed in order.
For each slice, ensure commands run end-to-end before moving to the next.
